JBPM-5801: name of element is not applied if you click on another node (fixed)

Christopher-Chianelli opened the pull request.

romartin commented:
ok to test

romartin commented:
Hey @pefernan , please, if possible, can you review this PR too? You're the forms master... :) 
Thanks!

hasys commented:
Jenkins execute full downstream build

pefernan commented:
Hi @Christopher-Chianelli I see what you try to fix but It could be done in different ways. Would you like to have a call to talk about that?

Christopher-Chianelli commented:
Greeting @pefernan, I would've preferred to done it in a different way too, but it seems the web browser doesn't update the fields in any classes until a defocus event. I tried firing a defocus event manually on the element, but to no avail. I am open to calls, talks and discussions.

hasys commented:
Jenkins execute full downstream build

pefernan commented:
@Christopher-Chianelli let's schedule a meeting to talk about that please. Ping me on IRC, my nickname is sr_pere.

@ederign @manstis please don't merge yet!

porcelli commented on kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-processing-engine/src/main/java/org/kie/workbench/common/forms/processing/engine/handling/impl/FormHandlerImpl.java:
```diff
@@ -250,6 +284,22 @@ public FormField findFormField(String fieldName) {
 
+    private native void getFields(Element element, ArrayList<InputElement> inputOutput,
+                                  ArrayList<TextAreaElement> textareaOutput) /*-{
+        var inputs = element.getElementsByTagName("input");
```
-1 for JSNI, JSInterop please!

Christopher-Chianelli commented:
Removed the ugly native JS code and replaced it with pretty JSInterop code. Gonna do some investigating to see if there is a way to cause fields to submit their values without accessing the elements.

pefernan commented:
@Christopher-Chianelli Monday I'll take a look.

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

pefernan commented on kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-processing-engine/src/main/java/org/kie/workbench/common/forms/processing/engine/handling/impl/FormHandlerImpl.java:
```diff
@@ -216,13 +224,40 @@ public void setReadOnly(boolean readOnly) {
 
+        //Update form fields
+        if (panel != null) {
+            for (Widget widget : panel){
```
@Christopher-Chianelli I think it could be done in a easier way than that.

FormHandler has a formFields list that contains all the fields on the form. You could iterate that list instead of iterating the elements on the widget. The FormField elements on that list are going to provide everything you neet to call the fieldChangeHandler.

The problem in your solution is that you're assuming that all the form elements are InputElements & TextAreas but they don't necessarily are going to be that. Form Engine support a lot of complex widgets and you're not processing them.

pefernan commented on kie-wb-common-forms/kie-wb-common-dynamic-forms/kie-wb-common-dynamic-forms-client/src/main/java/org/kie/workbench/common/forms/dynamic/client/DynamicFormRenderer.java:
```diff
@@ -222,6 +228,10 @@ public void unBind() {
         }
     }
 
+    public void setPanel(Panel panel) {
```
Avoid use GWT client classes on presenters. If you see my next message you'll see that there's no need to use this Panel

pefernan commented on kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-processing-engine/src/main/java/org/kie/workbench/common/forms/processing/engine/handling/FormHandler.java:
```diff
@@ -113,4 +114,9 @@ void addFieldChangeHandler(String fieldName,
+    /**
+     * Sets the panel (needed to get input elements before panel change)
+     */
+    void setPanel(Panel panel);
```
Same here

Christopher-Chianelli commented:
@porcelli, @pefernan, I managed to use formFields to get the HasValue widgets, and update the fields via the HasValue.getValue() method. I noticed some errors in the log related to updating the canvas due to undefined values, but the values seem to be updated fine from the user view. Some advanced widgets won't get there value changed, like when creating new rows in a table or editing it (this is using your recommended method, @pefernan ).

pefernan commented:
@Christopher-Chianelli let's sync up tomorrow

manstis commented:
Doesn't this need a QE review (for Forms) too?

triceo commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormPropertiesWidget.java:
```diff
@@ -286,6 +286,9 @@ private void doOpenSession(final ClientSession session) {
     }
 
     private void doClear() {
+        if (true || formRenderer.isValid()) {
```
Isn't this always `true`?

Christopher-Chianelli replied:
Yeah, I'm forcing a condition; currently developing with @pefernan. His version is working, but when I implemented his changes (this commit), nothing happens. I'm forcing it so I know the sync happens.

hasys commented:
Hi @manstis, @Christopher-Chianelli, @pefernan,

I just saw that this PR is still in progress and not ready to review. Please ping me if it's not valid any more and PR is ready to go.

Christopher-Chianelli commented:
@pefernan, @romartin, @jsoltes, can you please do a code review? Me and @pefernan managed to fix the issue for simple widgets such as Text Boxes and Text Areas, but more complex widgets such as Combo Boxes still have the issue.

pefernan commented on kie-wb-common-forms/kie-wb-common-dynamic-forms/kie-wb-common-dynamic-forms-client/src/main/java/org/kie/workbench/common/forms/dynamic/client/DynamicFormRendererViewImpl.java:
```diff
@@ -54,7 +60,8 @@ public void render(FormRenderingContext context) {
 
         if (context != null) {
-            formContent.add(layoutGenerator.buildLayout(context));
+            Panel panel = layoutGenerator.buildLayout(context);
```
@Christopher-Chianelli that's not a problem but maybe I'm forgetting something about this change, is there any reason for splitting this lines of code?

Christopher-Chianelli replied:
That is an artifact  from an earlier commit, where I passed Panel to the FormHandler. Since I am not passing the panel anymore to the FormHandler, it can be reverted.

pefernan commented on kie-wb-common-forms/kie-wb-common-dynamic-forms/kie-wb-common-dynamic-forms-client/src/main/java/org/kie/workbench/common/forms/dynamic/client/rendering/renderers/relations/multipleSubform/MultipleSubFormWidget.java:
```diff
@@ -250,6 +266,12 @@ public void deleteInstance(int index) {
 
+    @Override
+    public void forceModelSynchronization()
+    {
```
@Christopher-Chianelli needs some code formatting.

pefernan commented on kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-processing-engine/src/main/java/org/kie/workbench/common/forms/processing/engine/handling/impl/FormHandlerImpl.java:
```diff
@@ -234,6 +240,16 @@ public void clear() {
 
+    @Override
+    public void forceModelSynchronization()
+    {
```
@Christopher-Chianelli needs some code formatting.

Christopher-Chianelli commented:
Reformatted code; remove an unnecessary change.

pefernan commented on kie-wb-common-forms/kie-wb-common-forms-commons/kie-wb-common-forms-processing-engine/src/main/java/org/kie/workbench/common/forms/processing/engine/handling/impl/FormHandlerImpl.java:
```diff
@@ -234,6 +241,16 @@ public void clear() {
 
+    @Override
+    public void forceModelSynchronization() {
+        formFields.stream()
```
@Christopher-Chianelli coud you please add a test to cover this method?

Christopher-Chianelli commented:
@pefernan Resolved conflicts and added tests for `forceModelSynchronization`.

pefernan commented:
@porcelli could you approve? your requested changes are no longer needed.

porcelli commented on kie-wb-common-forms/kie-wb-common-dynamic-forms/kie-wb-common-dynamic-forms-client/src/main/java/org/kie/workbench/common/forms/dynamic/client/DynamicFormRenderer.java:
```diff
@@ -238,6 +238,13 @@ public boolean isValid() {
     }
 
+    public void forceModelSynchronization()
+    {
```
wrong code formatting?

porcelli commented:
@pefernan really small change request

Christopher-Chianelli commented:
@porcelli @hasys Fixed formatting and removed unused imports.

manstis commented:
@triceo @hasys Need your approval please..

manstis commented:
Jenkins please retest this.

romartin commented:
Jenkins please retest this

Christopher-Chianelli commented:
Jenkins execute full downstream build

manstis commented:
@Christopher-Chianelli You really need to see some of these ```kie-wb-common``` PRs through to completion!

Christopher-Chianelli commented:
@manstis, this one is finished (for a long time, but it seemed to be forgotten), but build been failing since the wildfly-dist property haven't been updated. Has the PR for updating wildfly-dist to "10.1.0.Final"? if so, I will run another downstream build.

Christopher-Chianelli commented:
Jenkins execute full downstream build

hasys commented:
Jenkins retest please

hasys commented:
Jenkins retest please

hasys commented:
Jenkins execute full downstream build

hasys commented:
Full downstream build is red, but it is due to some random fails in unrelated test. Binaries are available so will check it later today.

hasys commented on kie-wb-common-forms/kie-wb-common-dynamic-forms/kie-wb-common-dynamic-forms-client/src/main/java/org/kie/workbench/common/forms/dynamic/client/DynamicFormRenderer.java:
```diff
@@ -82,43 +79,26 @@ protected void init() {
-        renderDefaultForm(model,
-                          RenderMode.EDIT_MODE,
-                          null);
+        renderDefaultForm(model, RenderMode.EDIT_MODE, null);
```
Eclipse formatter shouldn't remove this new lines. Are you using the latest version?

hasys replied:
Ok, no problem. I guess since it is not violate our styles we can merge it, but it will be good to use unified KIE formatter in the future to eliminate excess formatting changes in PRs. 

@manstis up to you is it OK for this PR.
